### PR TITLE
split unit tests

### PR DESCRIPTION
**Pull request by brad-decker**
## Explanation
@FrederikBolding recently enabled parallelism in our E2E test suites, this pr does the same for our unit tests. The changes are enumerated below for informational purposes and to aid review. 

### CircleCI Config changes
The original 'test-unit' job did a handful of things:
1. Ran mocha tests
2. Ran jest tests on the development folder
3. Ran jest tests across the rest of the application
4. Validated jest unit test coverage
5. Ran jest-it-up and failed if coverage had increased by a margin of 5

To simplify the process of test splitting I first split this job into a handful of others:
1. test-unit-mocha now runs mocha tests and nothing else, this job will run concurrently with the test-unit-jest job and not impact the overall CI time
2. test-unit-jest-development now runs jest tests on the development folder using the development jest config.
3. test-unit-jest-main now runs unit tests using the main jest config and uses jest's shard feature to split tests across N processes (specified in circle config as parallelism -- currently 12). Note that the command overrides coverage reporters and targets (more on this later).
4. test-unit-jest-validate-coverage now combines the coverage reports from the processes above and validates all of the coverage from the jest.config.js file targets.
7. jest-it-up now runs jest-it-up because combining coverage before running it is necessary.

### merge-coverage
jest.config.js has a 'coverageThresholds' setting which previously was sufficient for validating we hit our goals. Unfortuantely with that setting each individual test run when split across multiple processes fails because none of them reach the global coverage targets. To combat this, the script run in the new test file override our jest.config.js settings for their own purposes. 

In merge-coverage, we use istanbul's libraries to combine coverage reports and create new ones, which gets us to a final coverage report state. It then uses our jest.config.js's coverageThresholds to validate we have hit our targets. You can review the merge-coverage.js file for specifics but it uses fast-glob to find files that match the glob specified for the target of the coverage threshold. Note that for some reason permissions/flask coverage is not reported even without my change and so I explicitly ignored it for purposes of coverage targets. 

### Before

image
image

### After

image
image

## Manual Testing Steps
1. You can visit the workflow for this branch and see all the runs and validate the success.

## Pre-merge author checklist

- [x] I've clearly explained:
  - [x] What problem this PR is solving
  - [x] How this problem was solved
  - [x] How reviewers can test my changes
- [ ] Sufficient automated test coverage has been added

### CODE REVIEW

File: test/merge-coverage.js [hunk 1]
```diff
@@ -0,0 +1,245 @@
+const fs = require('fs');
```
Comment (mcmire) [inline test/merge-coverage.js:1]: If I'm understanding this correctly, does this combine coverage reports from multiple distinct runs of `jest` into one report? If so, this is super rad and I think we could find this immensely useful on the monorepo side.

Comment (brad-decker) [reply test/merge-coverage.js:1]: Yep it sure does.

Comment (brad-decker) [reply test/merge-coverage.js:1]: and mocha :D cause they both use nyc

File: test/run-unit-tests.js [hunk 2]
```diff
@@ -0,0 +1,192 @@
...
+
+  const circleNodeIndex = parseInt(CIRCLE_NODE_INDEX ?? '0', 10);
+  const circleNodeTotal = parseInt(CIRCLE_NODE_TOTAL ?? '1', 10);
+
+  const maxProcesses = fakeParallelism > 0 ? fakeParallelism : circleNodeTotal;
+  const currentProcess = circleNodeIndex;
+
+  if (fakeParallelism) {
+    console.log(
+      `Using fake paralleism of ${fakeParallelism}. Your machine may become as useful as a brick during this operation.`,
```
Comment (mcmire) [inline test/run-unit-tests.js:153]: 😂

File: test/merge-coverage.js [hunk 3]
```diff
@@ -0,0 +1,245 @@
...
+ * @param {import('istanbul-lib-coverage').CoverageSummaryData} actual -
+ *  istanbul coverage summary representing actual coverage
+ */
+function checkCoverage(name, target, actual) {
+  const breakdown = generateConsoleReport(target, actual);
+  if (isCoverageInsufficient(target, actual)) {
+    const errorMsg = `Coverage thresholds for ${name} NOT met\n${breakdown}`;
+    throw new Error(errorMsg);
+  } else if (shouldCoverageBeBumped(target, actual)) {
+    const errorMsg = `Coverage thresholds for ${name} EXCEED and need bumped\n${breakdown}`;
```
Comment (legobeat) [inline test/merge-coverage.js:173]:
```suggestion
    const errorMsg = `Coverage thresholds for ${name} EXCEEDed and need bumping\n${breakdown}`;
```

Comment (brad-decker) [reply test/merge-coverage.js:173]: Agree that the wording is off on my take, but I would prefer something like 

```suggestion
    const errorMsg = `Coverage EXCEEDS threshold for ${name} and must be bumped\n${breakdown}`;
```

Comment (legobeat) [reply test/merge-coverage.js:173]: Agreed yours is better!

**Review by mcmire — COMMENTED** [review]
Looks good! Just one comment

File: .codecov.js [hunk 4]
```diff
@@ -0,0 +1,20 @@
+// Codecov uses a yaml file for its configuration and it targets line coverage.
```
[the remark this answers is not in the document]
Comment (brad-decker) [reply .codecov.js:1]: I'm fine with that.

[the remark this answers is not in the document]
Comment (brad-decker) [reply .codecov.js]: fixed in latest. Thanks @mcmire

**Review by mcmire — APPROVED** [review]
LGTM!

**Review by danjm — APPROVED** [review]
LGTM

**Pull request merged**